[WIP] WORKERFS prototype

Kagami opened the pull request:
As discussed earlier in #3641, I implemented some basic prototype of `WORKERFS`. Creating PR just to be sure I'm going in the right direction and to ask some questions.

* Are you ok with the API? I think people may want to work with multiple Blobs/Files at the same time and it will be convenient to make them available inside application with single mount call.
* I haven't implemented `setattr` and `llseek` yet. Is it ok to throw EPERM for all other calls? Or it's better to implement part of them in MEMFS-way too (rename, mkdir, unlink)?
* Also it may be worth to add some validation to `mount` (slashes in names, etc) and `read` (ranges, overlapping). Are you ok with the rest?
* I will add few other tests (files, unsupported calls, etc). Will it be enough?
* What about docs?

Thanks for your help and hints. Also, while digging through the code, I noticed that it's possible to provide device with your own read/write implementations. So all of this should be possible even without FS. But available by default FS is much more convenient to use of course.

Kagami commented:
Also, I just noticed. FS options should be quoted in order to prevent Closure to minify them. `root` option of NODEFS isn't quoted but it seems to be some whitelisted attribute so it works, but actually it shouldn't.

kripken commented:
Great start!

I'm curious how this could be done with a device, instead of FS - how would usage look like?

 * See the previous question, but your current API looks reasonable to me. A single mount call for multiple files seems fine. Side note, I predict that a good use case here would be to mount a single big file, and have compiled C++ code that can read parts out of it, like a zip file with its own directory structure, etc. - basically letting a whole package of files be accessed synchronously.
 * I am ok with EPERM for all the other stuff. The main use case is a simple read-only filesystem that just supports simple random-access reads, anything else might be nice but seems less important to me.
 * Validating the input to `mount` sounds good.
 * That initial test is a good start. Probably not much more is needed actually.
  * One thing I would like to see tested is to receive a `Blob` from the main thread, which for testing purposes could be just a simple Blob but in normal usage might be from a user opening a local file, or loading a Blob from IndexedDB. Perhaps we should even have an API that makes it easy? (just a random thought, no idea how that could work)
 * Some docs in `site/source/docs/api_reference/Filesystem-API.rst` would be good. We mention IDBFS and NODEFS there, for example.

kripken commented on tests/test_browser.py:
```diff
@@ -1007,6 +1007,18 @@ def test_fs_memfs_fsync(self):
+      Module.preRun = function() {
+        var blob = new Blob(['%s']);
+        FS.mkdir('/work');
+        FS.mount(WORKERFS, {blobs: [{name: 'blob.txt', data: blob}]}, '/work');
```
space after `{` and before `}`

kripken commented on src/library_workerfs.js:
```diff
@@ -0,0 +1,95 @@
+      if (parent) {
+        node.size = contents.size;
+        node.contents = contents;
+        node.reader = new FileReaderSync();
```
I suppose there is no reason not to make one `reader` for all of WORKERFS?

kripken commented on src/library_workerfs.js:
```diff
@@ -0,0 +1,95 @@
+    mount: function (mount) {
+      assert(ENVIRONMENT_IS_WORKER);
+      var root = WORKERFS.createNode(null, '/', WORKERFS.DIR_MODE, 0);
+      (mount.opts.files || []).forEach(function(file) {
```
What purpose does `files` provide? Isn't `blobs` all we want?

Kagami replied:
`File` objects are provided by e.g. `<input type=file>`. They are much handier in use because `File` instance incorporate name, data and timestamp. Also, we can also accept `FileList` here with `Array.prototype.forEach.call(...)`. What do you think?

kripken replied:
Oh, I see, so if a `File` is provided, we would have the name already? I think though that people might want a different name when adding it to the emscripten filesystem (the local file could be `/home/name/../../long-name` and the emscripten FS one might be `short-name`). Also I'm not sure the timestamp would be useful.

It seems simpler to me to just accept `Blobs`, unless I am missing an advantage here.

kripken replied:
I'm not opposed to it though, if it's documented and tested.

Kagami replied:
>I think though that people might want a different name

Then it's just `blobs: [{name: "short-name", data: file}]` since `File` is a `Blob`.

>Also I'm not sure the timestamp would be useful.

Some applications might use the mtime. Why not?

>unless I am missing an advantage here

I just thought it would be convenient to get file (or even files) from `<input type=file>` and pass it to WORKERFS right away.

>It seems simpler to me to just accept Blobs

Probably the only advantage is: users won't be confused by two different options?

>if it's documented and tested

It's tested by `test_fs_workerfs_read` and there are some docs. I can add more if it's not enough.

Kagami replied:
Also, File's name property [always contain only basename](https://developer.mozilla.org/en-US/docs/Web/API/File/name). There is no directories.

kripken replied:
Ok, sounds fine to me.

kripken commented on src/library_workerfs.js:
```diff
@@ -0,0 +1,95 @@
+      node.node_ops = WORKERFS.node_ops;
+      node.stream_ops = WORKERFS.stream_ops;
+      node.timestamp = (mtime || new Date).getTime();
+      if (parent) {
```
I'm not sure, but as a file under a mount, wouldn't they all have a parent?

Kagami replied:
Root node have `null` parent. See line 8.

Kagami commented:
(I will commit incrementally while working on this in order to simplify changes tracking.)

Kagami commented:
So far, I have:

* Fixed some review comments, answered to the rest
* Expanded test case, added lseek, setattr implementations
* Provided some docs

>One thing I would like to see tested is to receive a Blob from the main thread, which for testing purposes could be just a simple Blob but in normal usage might be from a user opening a local file

We have tests for constructed on the fly `Blob` and `File` objects but unfortunately we can't access to the real filesystem from JS unless some E2E testing framework like Selenium is used. I think there should be no real difference here?

>Validating the input to mount sounds good

I thought over this one more time and now I'm kind of confused: what level of validation do we actually want? Just `name.indexOf('/') === -1` or something more complex? Also I looked through other FS implementations and it turned out they provide only very basic input validation. What kind of logic will be more natural for Emscripten here? We can also enforce much stronger validation in other FS functions if needed.

>I'm curious how this could be done with a device, instead of FS - how would usage look like?

If we have e.g. blob that we want to map to VFS, something like this should be sufficient:

```js
var blob = new Blob([...]);
Module.preRun = function() {
  var id = FS.makedev(64, 0);
  FS.registerDevice(id, {
    read: function() { blob.slice(...) etc; },
  });
  FS.mkdev('/blob.txt', id);
};
```

kripken commented:
* We can leave the main thread issue for later, I suppose. I have some ideas for APIs for that, but they need more thought.
 * The same validation as other FSes have seems fine - no more and no less.
 * `FS.registerDevice` might actually be ok, if our focus is on single files anyhow, which I think it is. We should not require people to call `makedev` with a number, though, but perhaps `FS.createBlobFile` (parallel to `createPreloadedFile`), where it automatically makes an id, calls register device, etc.? What do you think?

Kagami commented:
>The same validation as other FSes have seems fine

Are you ok with the current state?

>if our focus is on single files anyhow, which I think it is

In my opinion there are valid use cases for multi-file WORKERFS. For example batch image converter or something like that. User selected multiple files with `<input type=file>` widget, JS pass it to WORKERFS via `files` option and they became available inside the application without any extra memory consumption. So it's almost transparent FS access.

Though I agree, `createBlobFile` might be useful in case of big-zipped-assets-pack (we probably have very different backgrounds, because I imagined photo-editing-batch-applications while doing this :smile:). Can we have both?

kripken commented:
Yes, definitely we should support both.

I guess I'm not sure which API feels more natural, both `FS.createBlobFile` (create a single blob-backed file) and `WORKERFS` (create a mount with possibly multiple blob-backed files). Do you have a preference? Also curious to hear from @inolen and @juj.

Kagami commented:
>Do you have a preference?

At first I imagined something like `FS.createBlobFile` and [even asked about it](https://github.com/kripken/emscripten/issues/3641#issuecomment-129597977), but WORKERFS should be quite handy in some use cases. For example it's possible to have multiple blob files with `FS.createBlobFile` by running it in a loop but it looks ugly IMO.

Mount interface is also consistent with existing file systems (NODEFS and IDBFS) and it should be easy to add to various Emscripten API wrappers (like [this](https://github.com/Kagami/ffmpeg.js#files) one) while `FS.createBlobFile` requires specialised code.

Kagami commented:
Also, since you said it's fine to have both APIs, can we merge this and implement `createBlobFile` later?
Are there something you don't like in this PR? I would like to test it with a real program and huge files (encoding with [ffmpeg](https://github.com/Kagami/ffmpeg.js)) and support for this FS in Emscripten will be very helpful.

Kagami commented:
Hm, or probably it's better if I build my own branch of Emscripten. I need to add one more entry to `emsdk_manifest.json` pointing to my repo&branch and run `emsdk install emscripten-workerfs-64bit && emsdk activate emscripten-workerfs-64bit`, right?

Kagami commented:
Also, one more thing, sorry for a lot of questions. Is it safe to use tagged `clang` (e.g. 1.34.4) with `emscripten-incoming`?

Kagami commented:
So I tested it in a real application. It works fine on a small files but still fails with "cannot enlarge memory arrays" on a huge files. I don't think that ffmpeg tries to read entire file into memory, probably it calls `mmap` and it compiles into sequential memory filling in Emscripten?
Also, it seems to be that second opening of file doesn't work properly. Will investigate further.

Kagami commented:
There was a bug in `read()` implementation actually.

Kagami commented:
So I tested it a bit more with ffmpeg and slightly increased TOTAL_MEMORY value. Everything works pretty well with 256M: I can successfully encode 600M+ files and overall memory consumption looks pretty decent (~8 times smaller than MEMFS in similar configuration).

kripken commented:
In general using the `incoming` branches on all 3 repos is always ok to do. I think `1.34.4` is the version on all of them.

I think this PR is good, but let's give the people I pinged before some more time to get a chance to provide feedback.

kripken commented on site/source/docs/api_reference/Filesystem-API.rst:
```diff
@@ -63,6 +62,14 @@ The *IDBFS* file system implements the :js:func:`FS.syncfs` interface, which whe
+
+.. note:: This file system is only for use when running code inside a worker.
+
+This file system provides read-only access to ``File`` and ``Blob`` objects inside a worker without copying entire data into memory and can potentially be used for a huge files.
```
"the entire data" and "for huge files" (add "the", remove "a")

kripken commented on src/library_workerfs.js:
```diff
@@ -0,0 +1,114 @@
+    reader: null,
+    mount: function (mount) {
+      assert(ENVIRONMENT_IS_WORKER);
+      WORKERFS.reader = new FileReaderSync();
```
multiple calls to `mount` will unnecessarily create the reader, can do `if (!WORKERFS.read) WORKERFS.reader = ..`

Kagami commented:
Fixed.

kripken commented:
Ok, this looks good, and since I see no complaints from anyone else, I think we can merge it. Please just squash the commits.

Kagami commented:
Done.

kripken commented:
Merged, thanks!
